### ISSUE TITLE
Extremely poor performance w/ large embedding matrices (sort of fixed, but plz see inside)

### ISSUE THREAD

**Issue by cbockman** [tensorflow/tensor2tensor#833]
Hi,

We have a t2t fork and recently went to the effort to pull our fork basically up-to-date with core repo master.

We then saw the run-time performance of our models plummet from, e.g., 30 steps/s ==> 1 step/s.

After much teeth gnashing, we tracked it down to an unfortunate interaction between a large (word) embedding matrix on the CPU and this update:

```
  def compute_gradients(self, loss, var_list=None, **kwargs):  # pylint: disable=arguments-differ
    gradients = self._opt.compute_gradients(loss, var_list, **kwargs)
    def cast_grad(g, v):
      if v is None or g is None:
        return (g, v)
      return (tf.cast(g, v.dtype), v)
    gradients = [cast_grad(g, v) for g, v in gradients]
    return gradients
```

Specifically, the version we were on was just

```
  def compute_gradients(self, loss, var_list=None, **kwargs):  # pylint: disable=arguments-differ
    return self._opt.compute_gradients(loss, var_list, **kwargs)
```

Given the code in block #1, what is happening makes sense.

The queries, then:

1) What is that new block of code used for?  E.g., is it some support peculiar to TPUs?  Are we safe to just (internally) comment it out and use the "old" version?  (Presumably, there is a very good reason this was changed!)

2) We've run into a couple of small other places that have to be handled very carefully to support word embeddings (at least if pinned on the CPU) (e.g., weight decay needs to be set to zero).

*If* it is safe to, in some circumstances, revert to the old compute_gradients code, is there any appetite for an internal flag (hparams.support_large_embeddings, or something like that--I'd need to put a little thought into where exactly this should live) that conditions the code paths appropriately?  

If this is something of interest, then would happily push a PR, to converge how we apparently need to manage things internally with t2t master.

**Comment by rsepassi** [comment]
Wow, that's unfortunate. Glad you were able to track down the change that degraded your performance. I believe a simple `if v.dtype == g.dtype: return (g, v)` would avoid the cast and the performance hit. I'll have that out shortly.

Please share other places in the code that do not play nicely with large embedding variables. I think we should try to have it supported out of the box without an additional hparam (unless necessity dictates otherwise).

And how large of an embedding variable are we talking about here?

**Comment by cbockman** [comment]
Thanks for the (very) quick response!

> And how large of an embedding variable are we talking about here?

Even just GloVe 6B 300d (https://nlp.stanford.edu/projects/glove/) makes things very sad.

> I believe a simple if v.dtype == g.dtype: return (g, v) would avoid the cast and the performance hit.

I think we can try this fix now--kind of surprising that tf.cast doesn't deal nicely with this case already?  Unless there are weird cases where tf.cast's x.dtype == dtype but that you'd still want to force a cast...

> Please share other places in the code that do not play nicely with large embedding variables. I think we should try to have it supported out of the box without an additional hparam (unless necessity dictates otherwise).

Main other place we ran into pain was weight_decay != 0 (https://github.com/tensorflow/tensor2tensor/blob/6a7ef7f79f56fdcb1b16ae76d7e61cb09033dc4f/tensor2tensor/layers/common_hparams.py#L67, which we inherited from as our starting point) crazily degrades things.  (Again, for understandable reasons, but took a bunch of hunting of course to id this as the culprit...)

Our internal "fix" is just to throw a (loud) warning if our internal embed_words modality is used, while this value is non-zero.

Otherwise, the pains of getting embeddings working mostly revolve around t2t and Estimator idiosyncrasies related to making sure that the embeddings aren't repeatedly loaded, etc.  

Have actually been meaning to get our word embedding modality up as a PR, but it is unfortunately a little hacky right now and dependent on various internal implementation details (we backend to GCS, for speed we pre-build the embeddings into a tensorflow model elsewhere and then load that model, etc.).  A publicly-shared word embedding modality probably deserves a bunch of more generalized code (inc. loading code) to support drop-in GloVe, etc...can see why you guys don't provide one right now.  :)

**Comment by epurdyf** [comment]
Just tried the suggested fix and it does not make the problem go away. So possibly the cast is necessary, but omitting it causes tensorflow to be smarter about not changing the gradients update to be non-sparse? It is not at all clear to me what is going on there.

**Comment by rsepassi** [comment]
That’s very odd since with that check it should be functionally equivalent
to how the code was before. Could you print out the dtypes and see if there
is a mismatch? If there is no mismatch, then what’s different from how the
code was before?

**Comment by cbockman** [comment]
Thanks @rsepassi , we'll try it first thing tomorrow and let you know.

**Comment by epurdyf** [comment]
Ahhhhhhh. In our case the values have dtype 'float32_ref' and the gradients have type 'float32':

``compute_gradients <dtype: 'float32_ref'> <dtype: 'float32'>``

Note that we found it to be much more efficient to pin the embeddings-related stuff to the CPU, which might be related to this.

**Comment by epurdyf** [comment]
Hm, this stack overflow question seems to imply that this is the expected behavior in general: https://stackoverflow.com/questions/37958706/in-tensorflow-what-is-the-difference-between-a-tensor-that-has-a-type-ending-in

**Comment by rsepassi** [comment]
Nice, that does seem like a good pattern.

The fix for the cast:
if v.dtype.base_dtype == g.dtype.base_dtype

**Comment by epurdyf** [comment]
OK, that does fix everything! Thank you so much!

**Comment by rsepassi** [comment]
Great. Please do think about cleaning up your code and submitting a PR to handle general pretrained embeddings.

**Comment by cbockman** [comment]
@rsepassi FYI this seems to be back; we're debugging & will share soln if(!)/when we surface, but am flagging in case anything jumps to mind.

Our last known good state was on our own fork, after applying your suggested fix (https://github.com/tensorflow/tensor2tensor/commit/abb03b120cd2b0b1dddec2e5aaa69e687ec2893b).
